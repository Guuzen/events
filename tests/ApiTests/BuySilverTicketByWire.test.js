"use strict";
const postgres = require('postgres');
const axios = require('axios').default;

const TestDatabase = require('./services/TestDatabase');
const testDatabaseConnection = postgres(require('./config/testDatabaseConfig'));
const testDatabase = new TestDatabase(testDatabaseConnection);

const EmailClient = require('./services/EmailClient');
const emailClient = new EmailClient(
    axios.create(require('./config/emailClientConfig'))
);

const eventsHttpClient = axios.create(require('./config/eventsClientConfig'));

const Manager = require('./actors/Manager');
const manager = new Manager(eventsHttpClient);

const Visitor = require('./actors/Visitor');
const visitor = new Visitor(eventsHttpClient, emailClient);

const emailTemplate = require('./matchTemplates/emailTemplate');
const responseTemplate = require('./matchTemplates/responseTemplate');
const responseSnapshotTemplate = require('./matchTemplates/responseSnapshotTemplate');

beforeAll(async () => {
    await testDatabase.prepare();
});

afterAll(async () => {
    await testDatabaseConnection.end();
});

jest.setTimeout(100 * 1000);

describe('Buy silver ticket by wire without promocode', function () {
    let eventId;
    test('manager creates event', async () => {
        const response = await manager.createEvent({
            name: '2019 foo event',
            domain: 'guuzen-events-nginx',
        });
        expect(response).toEqual(
            responseTemplate({
                id: expect.any(String)
            })
        );
        eventId = response.data.data.id;
    });

    test('manager see event in list', async () => {
        const response = await manager.getEventsInList();
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String)
                }
            ])
        );
    });

    test('manager see event by id', async () => {
        const response = await manager.getEventById(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String)
            })
        );
    });

    let tariffId;
    test('manager creates tariff', async () => {
        const response = await manager.createTariff({
            eventId: eventId,
            tariffType: 'silver_pass',
            segments: [
                {
                    price: {amount: 200, currency: 'RUB'},
                    term: {start: '2000-01-01 12:00:00', end: '3000-01-01 12:00:00'},
                }
            ]
        });
        expect(response).toEqual(
            responseTemplate({
                id: expect.any(String)
            })
        );
        tariffId = response.data.data.id;
    });

    test('manager see tariff in list', async () => {
        const response = await manager.getTariffsList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String)
                }
            ])
        );
    });

    test('manager see tariff by id', async () => {
        const response = await manager.getTariffById(tariffId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String)
            })
        );
    });

    let ticketId;
    test('manager creates ticket', async () => {
        const response = await manager.createTicket({
            eventId: eventId,
            tariffId: tariffId,
            number: '10002000'
        });
        expect(response).toEqual(
            responseTemplate({
                id: expect.any(String)
            })
        );
        ticketId = response.data.data.id;
    });

    test('manager see not reserved not deilivered ticket in list', async () => {
        const response = await manager.getTicketsList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String),
                    eventId: expect.any(String),
                    createdAt: expect.any(String),
                    reserved: false,
                    deliveredAt: null
                }
            ])
        );
    });

    test('manager see not reserved not delivered ticket by id', async () => {
        const response = await manager.getTicketById(ticketId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String),
                eventId: expect.any(String),
                createdAt: expect.any(String),
                reserved: false,
                deliveredAt: null
            })
        );
    });

    let orderId;
    test('visitor places order', async () => {
        const response = await visitor.placeOrder({
            tariffId: tariffId,
            firstName: 'John',
            lastName: 'Doe',
            email: 'john@example.com',
            phone: '+123456789'
        });
        expect(response).toEqual(
            responseTemplate({
                id: expect.any(String)
            })
        );
        orderId = response.data.data.id;
    });

    test('manager see not paid and not delivered order in list', async () => {
        const response = await manager.getOrdersList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String),
                    eventId: expect.any(String),
                    makedAt: expect.any(String),
                    productId: expect.any(String),
                    tariffId: expect.any(String),
                    userId: expect.any(String),
                    paid: false,
                    deliveredAt: null
                }
            ])
        );
    });

    test('manager see not paid and not delivered order by id', async () => {
        const response = await manager.getOrderById(orderId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String),
                eventId: expect.any(String),
                makedAt: expect.any(String),
                productId: expect.any(String),
                tariffId: expect.any(String),
                userId: expect.any(String),
                paid: false,
                deliveredAt: null
            })
        );
    });

    test('manager see reserved not delivered ticket in list', async () => {
        const response = await manager.getTicketsList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String),
                    eventId: expect.any(String),
                    createdAt: expect.any(String),
                    reserved: true,
                    deliveredAt: null,
                }
            ])
        );
    });

    test('manager see reserved not delivered ticket by id', async () => {
        const response = await manager.getTicketById(ticketId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String),
                eventId: expect.any(String),
                createdAt: expect.any(String),
                reserved: true,
                deliveredAt: null,
            })
        );
    });

    test('manager mark order paid', async () => {
        const response = await manager.markOrderPaid({
            orderId: orderId,
        });
        expect(response).toEqual(
            responseTemplate([])
        );
    });

    test('visitor receives email with ticket', async () => {
        const response = await visitor.getReceivedEmails();
        expect(response).toMatchSnapshot({
            items: [
                emailTemplate(),
            ]
        });
    });

    test('manager see paid and delivered order in list', async () => {
        const response = await manager.getOrdersList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String),
                    eventId: expect.any(String),
                    makedAt: expect.any(String),
                    productId: expect.any(String),
                    tariffId: expect.any(String),
                    userId: expect.any(String),
                    paid: true,
                    deliveredAt: expect.any(String),
                }
            ])
        );
    });

    test('manager see paid and delivered order by id', async () => {
        const response = await manager.getOrderById(orderId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String),
                eventId: expect.any(String),
                makedAt: expect.any(String),
                productId: expect.any(String),
                tariffId: expect.any(String),
                userId: expect.any(String),
                paid: true,
                deliveredAt: expect.any(String),
            })
        );
    });

    test('manager see reserved and delivered ticket in list', async () => {
        const response = await manager.getTicketsList(eventId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate([
                {
                    id: expect.any(String),
                    eventId: expect.any(String),
                    createdAt: expect.any(String),
                    reserved: true,
                    deliveredAt: expect.any(String),
                }
            ])
        );
    });

    test('manager see reserved and delivered ticket by id', async () => {
        const response = await manager.getTicketById(ticketId);
        expect(response).toMatchSnapshot(
            responseSnapshotTemplate({
                id: expect.any(String),
                eventId: expect.any(String),
                createdAt: expect.any(String),
                reserved: true,
                deliveredAt: expect.any(String),
            })
        );
    });
});
