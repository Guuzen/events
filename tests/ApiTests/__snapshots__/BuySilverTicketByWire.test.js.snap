// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Buy silver ticket by wire without promocode manager see event by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "domain": "guuzen-events-nginx",
      "id": Any<String>,
      "name": "2019 foo event",
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see event in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "domain": "guuzen-events-nginx",
        "id": Any<String>,
        "name": "2019 foo event",
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see not paid and not delivered order by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "cancelled": false,
      "currency": "RUB",
      "deliveredAt": null,
      "email": "john@example.com",
      "eventId": Any<String>,
      "firstName": "John",
      "id": Any<String>,
      "lastName": "Doe",
      "makedAt": Any<String>,
      "paid": false,
      "phone": "+123456789",
      "product": "silver_pass",
      "productId": Any<String>,
      "sum": "200",
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see not paid and not delivered order in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "cancelled": false,
        "currency": "RUB",
        "deliveredAt": null,
        "email": "john@example.com",
        "eventId": Any<String>,
        "firstName": "John",
        "id": Any<String>,
        "lastName": "Doe",
        "makedAt": Any<String>,
        "paid": false,
        "phone": "+123456789",
        "product": "silver_pass",
        "productId": Any<String>,
        "sum": "200",
        "tariffId": Any<String>,
        "userId": Any<String>,
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see not reserved not deilivered ticket in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "createdAt": Any<String>,
        "deliveredAt": null,
        "eventId": Any<String>,
        "id": Any<String>,
        "number": "10002000",
        "reserved": false,
        "type": "silver_pass",
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see not reserved not delivered ticket by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "createdAt": Any<String>,
      "deliveredAt": null,
      "eventId": Any<String>,
      "id": Any<String>,
      "number": "10002000",
      "reserved": false,
      "type": "silver_pass",
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see paid and delivered order by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "cancelled": false,
      "currency": "RUB",
      "deliveredAt": Any<String>,
      "email": "john@example.com",
      "eventId": Any<String>,
      "firstName": "John",
      "id": Any<String>,
      "lastName": "Doe",
      "makedAt": Any<String>,
      "paid": true,
      "phone": "+123456789",
      "product": "silver_pass",
      "productId": Any<String>,
      "sum": "200",
      "tariffId": Any<String>,
      "userId": Any<String>,
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see paid and delivered order in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "cancelled": false,
        "currency": "RUB",
        "deliveredAt": Any<String>,
        "email": "john@example.com",
        "eventId": Any<String>,
        "firstName": "John",
        "id": Any<String>,
        "lastName": "Doe",
        "makedAt": Any<String>,
        "paid": true,
        "phone": "+123456789",
        "product": "silver_pass",
        "productId": Any<String>,
        "sum": "200",
        "tariffId": Any<String>,
        "userId": Any<String>,
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see reserved and delivered ticket by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "createdAt": Any<String>,
      "deliveredAt": Any<String>,
      "eventId": Any<String>,
      "id": Any<String>,
      "number": "10002000",
      "reserved": true,
      "type": "silver_pass",
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see reserved and delivered ticket in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "createdAt": Any<String>,
        "deliveredAt": Any<String>,
        "eventId": Any<String>,
        "id": Any<String>,
        "number": "10002000",
        "reserved": true,
        "type": "silver_pass",
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see reserved not delivered ticket by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "createdAt": Any<String>,
      "deliveredAt": null,
      "eventId": Any<String>,
      "id": Any<String>,
      "number": "10002000",
      "reserved": true,
      "type": "silver_pass",
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see reserved not delivered ticket in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "createdAt": Any<String>,
        "deliveredAt": null,
        "eventId": Any<String>,
        "id": Any<String>,
        "number": "10002000",
        "reserved": true,
        "type": "silver_pass",
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see tariff by id 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Object {
      "id": Any<String>,
      "segments": Array [
        Object {
          "price": Object {
            "amount": "200",
            "currency": "RUB",
          },
          "term": Object {
            "end": "3000-01-01 12:00:00",
            "start": "2000-01-01 12:00:00",
          },
        },
      ],
      "tariffType": "silver_pass",
    },
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode manager see tariff in list 1`] = `
Object {
  "config": Anything,
  "data": Object {
    "data": Array [
      Object {
        "id": Any<String>,
        "segments": Array [
          Object {
            "price": Object {
              "amount": "200",
              "currency": "RUB",
            },
            "term": Object {
              "end": "3000-01-01 12:00:00",
              "start": "2000-01-01 12:00:00",
            },
          },
        ],
        "tariffType": "silver_pass",
      },
    ],
  },
  "headers": Anything,
  "request": Anything,
  "status": 200,
  "statusText": Anything,
}
`;

exports[`Buy silver ticket by wire without promocode visitor receives email with ticket 1`] = `
Object {
  "count": 1,
  "items": Array [
    Object {
      "Content": Object {
        "Body": "ticket number is 10002000",
        "Headers": Object {
          "Content-Transfer-Encoding": Array [
            "quoted-printable",
          ],
          "Content-Type": Array [
            "text/plain; charset=utf-8",
          ],
          "Date": Anything,
          "From": Array [
            "no-reply@event.com",
          ],
          "MIME-Version": Array [
            "1.0",
          ],
          "Message-ID": Anything,
          "Received": Anything,
          "Return-Path": Array [
            "<no-reply@event.com>",
          ],
          "Subject": Array [
            "Thanks for buy ticket",
          ],
          "To": Array [
            "john@example.com",
          ],
        },
        "MIME": null,
        "Size": 315,
      },
      "Created": Any<String>,
      "From": Object {
        "Domain": "event.com",
        "Mailbox": "no-reply",
        "Params": "",
        "Relays": null,
      },
      "ID": Any<String>,
      "MIME": null,
      "Raw": Object {
        "Data": Any<String>,
        "From": "no-reply@event.com",
        "Helo": "guuzen-events-nginx",
        "To": Array [
          "john@example.com",
        ],
      },
      "To": Array [
        Object {
          "Domain": "example.com",
          "Mailbox": "john",
          "Params": "",
          "Relays": null,
        },
      ],
    },
  ],
  "start": 0,
  "total": 1,
}
`;
